samples/crypto/mbedtls_benchmark/src/benchmark.c

```diff
@@ -0,0 +1,1068 @@
+ * purpose implementing necessary functions.
+ */
+
+volatile int mbedtls_timing_alarmed;
+
+static struct k_delayed_work mbedtls_alarm;
+static void mbedtls_alarm_timeout(struct k_work *work);
+
+static void mbedtls_alarm_timeout(struct k_work *work)
+{
+	mbedtls_timing_alarmed = 1;
+}
+
+void mbedtls_set_alarm(int seconds)
+{
+	mbedtls_timing_alarmed = 0;
+
+	k_delayed_work_submit(&mbedtls_alarm, K_SECONDS(seconds));
+}
+
+/*
+ * For heap usage estimates, we need an estimate of the overhead per allocated
+ * block. ptmalloc2/3 (used in gnu libc for instance) uses 2 size_t per block,
+ * so use that as our baseline.
+ */
+#define MEM_BLOCK_OVERHEAD  (2 * sizeof(size_t))
+
+#define BUFSIZE         1024
+#define HEADER_FORMAT   "  %-24s :  "
+#define TITLE_LEN       25
+
+#define OPTIONS                                                    \
+	"md4, md5, ripemd160, sha1, sha256, sha512,\n"             \
+	"arc4, des3, des, camellia, blowfish, chacha20,\n"         \
+	"aes_cbc, aes_gcm, aes_ccm, aes_ctx, chachapoly,\n"        \
+	"aes_cmac, des3_cmac, poly1305,\n"                         \
+	"havege, ctr_drbg, hmac_drbg,\n"                           \
+	"rsa, dhm, ecdsa, ecdh.\n"
+
+#if defined(MBEDTLS_ERROR_C)
+#define PRINT_ERROR {                                            \
+	mbedtls_strerror(ret, (char *)tmp, sizeof(tmp));         \
+	mbedtls_printf("FAILED: %s\n", tmp);                     \
+	}
+#else
+#define PRINT_ERROR {                                            \
+	mbedtls_printf("FAILED: -0x%04x\n", -ret);               \
+	}
+#endif
+
+#define TIME_AND_TSC(TITLE, CODE)                                     \
+do {                                                                  \
+	unsigned long ii, jj;                                         \
+	u32_t tsc;                                                    \
+	int ret = 0;                                                  \
+								      \
+	mbedtls_printf(HEADER_FORMAT, TITLE);                         \
+								      \
+	mbedtls_set_alarm(1);                                         \
+	for (ii = 1; ret == 0 && !mbedtls_timing_alarmed; ii++) {     \
+		ret = CODE;                                           \
+	}                                                             \
+								      \
+	tsc = k_cycle_get_32();                                       \
+	for (jj = 0; ret == 0 && jj < 1024; jj++) {                   \
+		ret = CODE;                                           \
+	}                                                             \
+								      \
+	if (ret != 0) {                                               \
+		PRINT_ERROR;                                          \
+	}                                                             \
+								      \
+	mbedtls_printf("%9lu KiB/s,  %9lu cycles/byte\n",             \
+		       ii * BUFSIZE / 1024,                           \
+		       (k_cycle_get_32() - tsc) / (jj * BUFSIZE));   \
```

Comment:
I'm not sure if `k_cycle_get_32` is  a good function here. For my understanding, we should verify here how many CPU cycles the crypto calculations took. On the other hand `k_cycle_get_32` returns, according to the docs:
```
unspecified units called cycles. The duration of a cycle is determined by the board hardware used by the kernel, and is typically measured in nanoseconds.
```
As for instance nRF uses RTC for HW clock, I get awkward results, like this:
```
AES-CBC-128              :        891 KiB/s,          0 cycles/byte
```
Unfortunately I'm no expert in this area, but maybe someone will have some suggestions for a better approach here. A quick solution that comes up to my mind is to use `SYS_CLOCK_HW_CYCLES_TO_NS` macro and provide result in ns/byte, which gives reasonable output on both nRF and qemu_x86.

Reply:
Ok